File: 2054-the-number-of-the-smallest-unoccupied-chair/2054-the-number-of-the-smallest-unoccupied-chair.cs
public class Solution {
    public int SmallestChair(int[][] times, int targetFriend) {
        PriorityQueue<int,int> pq = new PriorityQueue<int,int>();


        Dictionary<int,int>mp = new Dictionary<int,int>();
        Dictionary<int,List<int>> mp2 = new Dictionary<int,List<int>>();

        int targetindex =times[targetFriend][0];
        for(int i = 0; i< times.Length; i++)
        {
           mp[times[i][0]]= times[i][1];
           
        }
        int curavchair = 0;
        for(int i = 1; i<= targetindex; i++)
        {
            if(mp2.ContainsKey(i) == true)
            {
                foreach(var x in  mp2[i])
                {
                    pq.Enqueue(x,x);
                }
            }
            if(mp.ContainsKey(i) == true)
            {
                if(pq.Count >0)
                {
                    if(i == targetindex)
                        return pq.Peek();
                    int end = mp[i];
                    if(mp2.ContainsKey(end))
                        mp2[end].Add(pq.Peek());
                    else
                        mp2.Add(end, new List<int>(){pq.Peek()});
                    pq.Dequeue();
                    
                }
                else
                {
                    if(i== targetindex)
                        return curavchair;
                    int end = mp[i];
                    if(mp2.ContainsKey(end))
                        mp2[end].Add(curavchair);
                    else
                        mp2.Add(end, new List<int>(){curavchair});
                    
                    curavchair++;
                    
                }
            }
            
        }
        return 0;
        

        
    }
}

File: 2054-the-number-of-the-smallest-unoccupied-chair/2054-the-number-of-the-smallest-unoccupied-chair.cpp
class Solution {
public:
    int smallestChair(vector<vector<int>>& times, int targetFriend) {
        
        unordered_map<int,int>mp;
        unordered_map<int,vector<int>>mp2;

        int targetindex =times[targetFriend][0];
        for(int i = 0; i< times.size(); i++)
        {
           mp[times[i][0]]= times[i][1];
           
        }
        int curavball = 0;
        priority_queue<int, vector<int>, greater<int>> pq;
        for(int i = 1; i<= targetindex; i++)
        {
            if(mp2.count(i)>0)
            {
                for(auto x: mp2[i])
                {
                    pq.push(x);
                }
            }
            if(mp.count(i)>0)
            {
                if(!pq.empty())
                {
                    if(i == targetindex)
                        return pq.top();
                    int end = mp[i];
                    mp2[end].push_back(pq.top());
                    pq.pop();
                    
                }
                else
                {
                     int end = mp[i];
                    mp2[end].push_back(curavball);
                    curavball++;
                    if(i== targetindex)
                        return curavball-1;
                }
            }
            
        }
        cout<<"**********"<<"\n";
        return 0;
        

    }
};